[ONNX] Fix bug in ONNXToTorch PadOp's pads tensor rearrangement

sagarkulkarni19 opened the pull request:
Fix the pad tensor rearrangement such that we change the representation from [x1_begin, x2_begin, ..., x1_end, x2_end,...] to [xn_begin, xn_end, ...., x2_begin, x2_end, x1_begin, x1_end] where x1, x2 .. xn are the dimensions of the pads tensor argument.

sagarkulkarni19 commented:
Fixes #3484

sagarkulkarni19 commented:
Can you please help me add reviewers? I cc'ed folks that have originally added the code here or added similar fixes. 
cc: @vivekkhandelwal1 @Shukla-Gaurav @rsuderman

vivekkhandelwal1 commented:
> Can you please help me add reviewers? I cc'ed folks that have originally added the code here or added similar fixes. cc: @vivekkhandelwal1 @Shukla-Gaurav @rsuderman

Done @sagarkulkarni19

sagarkulkarni19 commented:
> > Can you please help me add reviewers? I cc'ed folks that have originally added the code here or added similar fixes. cc: @vivekkhandelwal1 @Shukla-Gaurav @rsuderman
> 
> Done @sagarkulkarni19

Thanks!

sagarkulkarni19 commented:
This patch fails e2e ONNX tests. I'm looking into it further.

zjgarvey commented:
@sagarkulkarni19 Can you please check out the PR <https://github.com/sagarkulkarni19/torch-mlir/pull/1> and then sync your branch with main? I need this important fix, so thanks for the work!

sagarkulkarni19 commented:
> @sagarkulkarni19 Can you please check out the PR [sagarkulkarni19#1](https://github.com/sagarkulkarni19/torch-mlir/pull/1) and then sync your branch with main? I need this important fix, so thanks for the work!

Thanks for debugging this further @zjgarvey ! How should I proceed with the commits-- would it be okay if I squashed all commits into the first one and added a `Co-authored-by:` and added you?

zjgarvey commented:
> > @sagarkulkarni19 Can you please check out the PR [sagarkulkarni19#1](https://github.com/sagarkulkarni19/torch-mlir/pull/1) and then sync your branch with main? I need this important fix, so thanks for the work!
> 
> Thanks for debugging this further @zjgarvey ! How should I proceed with the commits-- would it be okay if I squashed all commits into the first one and added a `Co-authored-by:` and added you?

When this pr gets merged it should squash automatically, so I wouldn't worry about it.

zjgarvey commented:
@sagarkulkarni19 would you mind syncing the branch with main? The recent onnx version update will cause some tests to crash without the updates to xfail_sets.py. 

You can do this from your branch homepage on github (with the sync fork button), or you can run the following from your local repo:

```
git remote add upstream https://github.com/llvm/torch-mlir.git
git fetch upstream
git merge upstream/main
```

check for conflicts, then push the merge commit.

sagarkulkarni19 commented:
Ah apologies, I forgot to rebase on top of main! @zjgarvey .Done now.

zjgarvey commented:
> Ah apologies, I forgot to rebase on top of main! @zjgarvey .Done now.

Thanks! I'm re-running the CI and if it runs into any issues, I'll help you debug.

zjgarvey commented:
Okay, it looks like something bad is going on, so I'll take a look.

zjgarvey commented on lib/Dialect/Torch/IR/TorchOps.cpp:
```diff
@@ -3664,7 +3664,9 @@ OpFoldResult AtenSliceTensorOp::fold(FoldAdaptor adaptor) {
 
-  // If the input and output shapes are the same we can just fold:
+  // If the input and output shapes are the same & step == 1 we can fold:
+  if (step.getValue().getSExtValue() != 1)
```
Ah, I think the issue is resolved with:

```suggestion
  if (!step || step.getValue().getSExtValue() != 1)
```

It was crashing because step failed the dyn cast on initial applications of the folder (this was happening on `ScaledDotProductAttentionDifferentModule_basic`).

Sorry for submitting the sloppy fix earlier!

sagarkulkarni19 replied:
Thanks for debugging the issue! Commited the fix.

zjgarvey commented:
@sagarkulkarni19 

Yay, it looks like some e2e tests xpassed!

Please delete the following lines:
<https://github.com/llvm/torch-mlir/blob/main/projects/pt1/e2e_testing/xfail_sets.py#L2219-L2220>

To remove the tests 
```python
    "FlipModuleStaticShape_basic",
    "FlipNegativeIndexModule_basic",
```
from `ONNX_XFAIL_SET`.

Thanks for the quick responses!

zjgarvey commented:
I checked out the other configs for those tests locally and it doesn't look like other changes need to be made. Once this passes the CI, do you mind if I merge it right away?

sagarkulkarni19 commented:
Oh haha, sorry I did not see that they `unexpectedly passed` ! This is great, thanks! 

>I checked out the other configs for those tests locally and it doesn't look like other changes need to be made. Once this passes the CI, do you mind if I merge it right away?

Sure! Let me push the fix right now.

sagarkulkarni19 commented:
I think it's all good now. Please run the workflows now! @zjgarvey and feel free to merge it.
